### PR TITLE
[hdf5] Update HDF5 to 1.12 

### LINKED ISSUE #11745: [hdf5] update to 1.12
Library name: hdf5

New version number: 1.12

Other information that may be useful (release notes, etc...)

### PR DESCRIPTION
**Pull request by Neumann-A**
and use gitlab live-clones/hdf5 instead of binary download

closes #11745

### CODE REVIEW

**Comment by cenit** [comment]
Personally, I am not a fan of unofficial urls.
If you want to download from repo, isn't it better to use vcpkg_from_git and point to https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse ?

File: ports/hdf5/portfile.cmake [hunk 1]
```diff
@@ -1,20 +1,12 @@
...
-    REF hdf5
-    PATCHES
-        hdf5_config.patch       
-        fix-generate.patch      # removes the build of static targets in shared builds
-        static-targets.patch    # maps the internal static tagets to the shared targets if building as a dynamic library
-        export-private.patch    # exports two additional functions in shared builds to make hl/tools/h5watch build in shared builds. 
+    REPO  live-clones/hdf5
+    REF hdf5-1_12_0
+    SHA512 d84df1ea72dc6fa038440a370e1b1ff523364474e7f214b967edc26d3191b2ef4fe1d9273c4a086a5945f1ad1ab6aa8dbcda495898e7967b2b73fd93dd5071e0
+    HEAD_REF develop
```
Comment (cenit) [inline ports/hdf5/portfile.cmake:8]: vcpkg_from_git(
    OUT_SOURCE_PATH SOURCE_PATH
    URL https://bitbucket.hdfgroup.org/scm/hdffv/hdf5.git
    REF 83385326ef1588167f2b37300313a6e51e51ffca
```

**Comment by Neumann-A** [comment]
@cenit: You actually tried the code?
```
Starting package 3/3: hdf5:x64-windows-static
Building package hdf5[core,szip,zlib]:x64-windows-static...
-- Fetching https://bitbucket.hdfgroup.org/scm/hdffv/hdf5.git...
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:72 (message):
    Command failed: "C:/Program Files/Git/mingw64/bin/git.exe" fetch https://bitbucket.hdfgroup.org/scm/hdffv/hdf5.git 83385326ef1588167f2b37300313a6e51e51ffca --depth 1 -n
    Working Directory: D:/vcpkg_common/downloads/git-tmp
    Error code: 128
    See logs for more information:
      D:\qt2\buildtrees\hdf5\git-fetch-x64-windows-static-err.log
```

```
fatal: expected shallow/unshallow, got The client mentioned a shallow object that is missing on the server. This may be due to a rebase or other forced update on the server. You should recreate your shallow clone.
fatal: the remote end hung up unexpectedly
```

**Comment by cenit** [comment]
Working on my Pc 
Anyway, feel free to do what you prefer. Just stating my opinion about unofficial repositories...

**Comment by Neumann-A** [comment]
@NancyLi1013: Can i assume that none of those regressions have anything to do with the changes made in this PR?

**Comment by NancyLi1013** [comment]
@Neumann-A 
Thanks for your kindly reminder. I will rerun this PR and confirm if those regression can be solved.
Currently, I'm not sure how the regression for `field3d `on x64-linux is caused.

```
/mnt/vcpkg-ci/buildtrees/field3d/src/v1.7.2-ae08f6d321/src/Field3DFileHDF5.cpp:1138:71: error: too few arguments to function ‘herr_t H5Oget_info_by_name3(hid_t, const char*, H5O_info2_t*, unsigned int, hid_t)’
   status = H5Oget_info_by_name(loc_id, itemName, &infobuf, H5P_DEFAULT);
                                                                       ^
In file included from /mnt/vcpkg-ci/install/x64-linux/include/H5Apublic.h:22:0,
                 from /mnt/vcpkg-ci/install/x64-linux/include/hdf5.h:23,
                 from /mnt/vcpkg-ci/buildtrees/field3d/src/v1.7.2-ae08f6d321/src/Field3DFileHDF5.cpp:50:
/mnt/vcpkg-ci/install/x64-linux/include/H5Opublic.h:188:15: note: declared here
 H5_DLL herr_t H5Oget_info_by_name3(hid_t loc_id, const char *name, H5O_info2_t *oinfo,
               ^~~~~~~~~~~~~~~~~~~~
/mnt/vcpkg-ci/buildtrees/field3d/src/v1.7.2-ae08f6d321/src/Field3DFileHDF5.cpp: In function ‘herr_t Field3D::v1_7::InputFileHDF5::parseLayers(hid_t, const char*, const H5L_info2_t*, void*)’:

   status = H5Oget_info_by_name (loc_id, itemName, &infobuf, H5P_DEFAULT);
```

**Comment by NancyLi1013** [comment]
@Neumann-A 
Seems the regression is related with this `hdf5`. 
Could you please look into this?

**Comment by Neumann-A** [comment]
@NancyLi1013: TODO for the future: Fix field3d hidden MPI dependency

**Comment by cenit** [comment]
> @NancyLi1013: TODO for the future: Fix field3d hidden MPI dependency

It’s already almost done in my opencv 4.3 pr

**Comment by JackBoosY** [comment]
@cenit So, should we waiting for your PR #11130?

File: ports/field3d/portfile.cmake [hunk 2]
```diff
@@ -11,16 +8,19 @@ endif()
...
         fix-build_error.patch
+        hdf5.api.patch # Switches the HDF5 default API for this port to 1.10
 )
 
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
+    OPTIONS
+        "-DINSTALL_DOCS:BOOL=OFF"
 )
```
Comment (NancyLi1013) [inline ports/field3d/portfile.cmake:25]: Could you please update this

```
file(COPY ${SOURCE_PATH}/COPYING ${SOURCE_PATH}/README DESTINATION ${CURRENT_PACKAGES_DIR}/share/field3d)
file(RENAME ${CURRENT_PACKAGES_DIR}/share/field3d/COPYING ${CURRENT_PACKAGES_DIR}/share/field3d/copyright)
```
as

`file(INSTALL ${SOURCE_PATH}/COPYING ${SOURCE_PATH}/README DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright )`?

File: ports/hdf5/portfile.cmake [hunk 3]
```diff
@@ -1,20 +1,13 @@
...
+vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
-    ARCHIVE ${ARCHIVE}
-    REF hdf5
-    PATCHES
-        hdf5_config.patch       
-        fix-generate.patch      # removes the build of static targets in shared builds
-        static-targets.patch    # maps the internal static tagets to the shared targets if building as a dynamic library
-        export-private.patch    # exports two additional functions in shared builds to make hl/tools/h5watch build in shared builds. 
+    REPO  live-clones/hdf5
```
Comment (NancyLi1013) [inline ports/hdf5/portfile.cmake:5]:
```suggestion
    REPO live-clones/hdf5
```
Could you please update this?

File: ports/hdf5/portfile.cmake [hunk 4]
```diff
@@ -1,20 +1,13 @@
...
+    REF hdf5-1_12_0
+    SHA512 d84df1ea72dc6fa038440a370e1b1ff523364474e7f214b967edc26d3191b2ef4fe1d9273c4a086a5945f1ad1ab6aa8dbcda495898e7967b2b73fd93dd5071e0
+    HEAD_REF develop
+    PATCHES hdf5_config.patch
```
Comment (NancyLi1013) [inline ports/hdf5/portfile.cmake:9]:
```suggestion
    PATCHES 
        hdf5_config.patch
```
Could you please update this?

File: ports/hdf5/portfile.cmake [hunk 5]
```diff
@@ -1,20 +1,13 @@
...
+            szip.patch
```
Comment (NancyLi1013) [inline ports/hdf5/portfile.cmake:10]:
```suggestion
       szip.patch
```

**Comment by cenit** [comment]
yes please. I just pushed the patch for field3d there.
Waiting for CI results, hoping for that PR to be done...

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
@cenit 
Thanks for your attention and kindly reminder. Actually, we always recommend to use official repos except for some specific reasons.  And now I noticed that @Neumann-A has updated as official repo for `hdf5` in this PR. Really appreciate.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
@Neumann-A 
Is there anything else to be done for this PR? If not, I will try to test these features.

**Comment by Neumann-A** [comment]
@NancyLi1013: Nothing to add. Feel free to resolve the conflict with the baseline.

**Comment by JackBoosY** [comment]
Can you merge the latest changes into this branch and use [this](https://github.com/microsoft/vcpkg/files/4768959/ci.baseline.txt) to replace `ci.baseline.txt`?

**Comment by Neumann-A** [comment]
@JackBoosY: I cleaned the baseline. If you need more changes than that those should probably be in another PR.

**Review by JackBoosY — APPROVED** [review]

**Comment by NancyLi1013** [comment]
All features have passed with the following triplets:
- x86-windows
- x64-windows
- x64-windows-static

Note: Feature `fortran ` has not been supported yet.

**Comment by Neumann-A** [comment]
@JackBoosY @NancyLi1013´: Is there something blocking the merge of this PR?

**Comment by JackBoosY** [comment]
@Neumann-A See https://github.com/microsoft/vcpkg/pull/11747#issuecomment-641746157

Waiting for merge #11130.

**Review by dan-shaw — APPROVED** [review]

**Comment by dan-shaw** [comment]
LGTM, just waiting for #11130 as suggested above. @Neumann-A Let us know if it makes more sense to merge this PR first.

**Comment by Neumann-A** [comment]
@dan-shaw: seems like #11130 is still not finished so merging this one first seems more reasonable

**Comment by cenit** [comment]
#11130 is finished...

**Comment by BillyONeal** [comment]
This change broke a lot of stuff :(

**Comment by Neumann-A** [comment]
@BillyONeal. Instead of patching the sources use the following in the CMakeLists.txt:
```
add_definitions(-UH5_USE_112_API_DEFAULT)
add_definitions(-DH5_USE_110_API_DEFAULT)
```
there is probably also a way to set those via an environment variable from the portfile or using `VCPKG_CXX_FLAGS`

**Pull request merged**